tee_supplicant/src/tee_supplicant.c

```diff
@@ -0,0 +1,296 @@
+{
+	if (idx >= num_params) {
+		return -1;
+	}
+
+	switch (params[idx].attr & TEE_PARAM_ATTR_TYPE_MASK) {
+	case TEE_PARAM_ATTR_TYPE_VALUE_INPUT:
+	case TEE_PARAM_ATTR_TYPE_VALUE_OUTPUT:
+	case TEE_PARAM_ATTR_TYPE_VALUE_INOUT:
+		*value = (void *)&params[idx].a;
+		return 0;
+	default:
+		return -1;
+	}
+}
+
+static struct suppl_shm *find_tshm(int id)
+{
+	struct suppl_shm *tshm = NULL;
+
+	tee_supp_mutex_lock(&shm_mutex);
+
+	tshm = shm_head;
+	while (tshm && tshm->id != id) {
+		tshm = tshm->next;
+	}
+
+	tee_supp_mutex_unlock(&shm_mutex);
+
+	return tshm;
+}
+
+/* Get parameter allocated by secure world */
+static int get_param(size_t num_params, struct tee_param *params,
+		     const uint32_t idx, TEEC_SharedMemory *shm)
+{
+	struct suppl_shm *tshm = NULL;
+	size_t offs = 0;
+	size_t sz = 0;
+
+	if (idx >= num_params) {
+		return -1;
+	}
+
+	switch (params[idx].attr & TEE_PARAM_ATTR_TYPE_MASK) {
+	case TEE_PARAM_ATTR_TYPE_MEMREF_INPUT:
+	case TEE_PARAM_ATTR_TYPE_MEMREF_OUTPUT:
+	case TEE_PARAM_ATTR_TYPE_MEMREF_INOUT:
+		break;
+	default:
+		return -2;
+	}
+
+	memset(shm, 0, sizeof(*shm));
+
+	tshm = find_tshm(MEMREF_SHM_ID(params + idx));
+	if (!tshm) {
+		/*
+		 * It doesn't make sense to query required size of an
+		 * input buffer.
+		 */
+		if ((params[idx].attr & TEE_PARAM_ATTR_TYPE_MASK) ==
+		    TEE_PARAM_ATTR_TYPE_MEMREF_INPUT) {
+			return -3;
+		}
+
+		/*
+		 * Buffer isn't found, the caller is querying required size
+		 * of the buffer.
+		 */
+		return 0;
+	}
+
+	sz = MEMREF_SIZE(params + idx);
+	offs = MEMREF_SHM_OFFS(params + idx);
+	if ((sz + offs) < sz) {
+		return -4;
+	}
+	if ((sz + offs) > tshm->size) {
+		return -5;
+	}
+
+	shm->flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT;
+	shm->size = sz;
+	shm->id = MEMREF_SHM_ID(params + idx);
+	shm->buffer = (uint8_t *)tshm->p + offs;
+
+	return 0;
+}
+
+static void uuid_from_octets(TEEC_UUID *d, const uint8_t s[TEE_UUID_LEN])
+{
+	d->timeLow = (s[0] << 24) | (s[1] << 16) | (s[2] << 8) | s[3];
+	d->timeMid = (s[4] << 8) | s[5];
+	d->timeHiAndVersion = (s[6] << 8) | s[7];
+	memcpy(d->clockSeqAndNode, s + 8, sizeof(d->clockSeqAndNode));
+}
+
+static int load_ta(uint32_t num_params, struct tee_param *params)
+{
+	int ta_found = 0;
+	size_t size = 0;
+	struct param_value *val_cmd = NULL;
+	TEEC_UUID uuid = { 0 };
+	TEEC_SharedMemory shm_ta = { 0 };
+
+	if (num_params != 2 || get_value(num_params, params, 0, &val_cmd) ||
+	    get_param(num_params, params, 1, &shm_ta)) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	uuid_from_octets(&uuid, (void *)val_cmd);
+
+	size = shm_ta.size;
+	ta_found = TEECI_LoadSecureModule(&uuid, shm_ta.buffer, &size);
+	if (ta_found != TA_BINARY_FOUND) {
+		LOG_ERR("TA not found");
+		return TEEC_ERROR_ITEM_NOT_FOUND;
+	}
+
+	MEMREF_SIZE(params + 1) = size;
+
+	/*
+	 * If a buffer wasn't provided, just tell which size it should be.
+	 * If it was provided but isn't big enough, report an error.
+	 */
+	if (shm_ta.buffer && size > shm_ta.size) {
```

Comment:
Should you do this before calling `TEECI_LoadSecureModule` ?

Reply:
No. Return code TA_BINARY_FOUND doesn't mean everything is ok. It means requested TA was found ans accessible. If size of is TA bigger then buffer TA will not be copied. `size` variable contains size of TA after `TEECI_LoadSecureModule()`